### DISCUSSION TITLE
How can one define the headers once for all endpoints?

### DISCUSSION THREAD

**Discussion by artisanphil** [danielgtaylor/huma#658] [Q&A] [answered]
Currently I'm defining headers like this:
```
type ResponseHeaders struct {
	CacheControl string `header:"Cache-Control" example:"max-age=3600" json:"-"`
}

type MultiplePropertyOutput struct {
	ResponseHeaders
	Body struct {
		Hits    int        `json:"hits" example:"100" doc:"Number of properties returned"`
		Results []Property `json:"results"`
		Links   []Link     `json:"links"`
		Status  Status     `json:"status"`
	} `json:"body"`
}
```

With this I have to add the response header struct to every output struct that I define. I would like to set the headers just once where I define the routes. Otherwise I have to do it like this for each route:

```
resp := &models.MultiplePropertyOutput{}
resp.CacheControl = "max-age=86400"
```

Is that possible? I also have a small side-issue with this in that it returns "responseheaders" as a header in addition to cache-control:

![image](https://github.com/user-attachments/assets/7479a99f-e6af-45fd-bc61-8c3f1b0fe63f)

Appreciate any help that I can get on this.

**Comment by danielgtaylor** [comment]
@artisanphil this might be a good case for a middleware. You can document it as part of the top-level API description saying it's a common header or loop through and add it to all the operations as needed. Example:

https://go.dev/play/p/ezWCcctZRg9

For the other issue, I believe #656 fixes is and will get released ASAP.

**Reply by artisanphil** [reply, marked as answer]
@danielgtaylor Thank you. 

I got it working like this:

```
        api.UseMiddleware(func(ctx huma.Context, next func(huma.Context)) {
		ctx.SetHeader("Cache-Control", "max-age=86400")
		next(ctx)
	})
```

For the documentation itself, I'm now just describing the headers on the overview page rather than per endpoint (e.g. under `config.Info.Description`).

**Reply by danielgtaylor** [reply]
Whoops, this is what happens when I rush 😊 I forgot the `next(ctx)` call in the middleware. Fixed example: https://go.dev/play/p/ezWCcctZRg9
